### ISSUE TITLE
Empty objects should be data, Far should be marked as such

### ISSUE THREAD

**Issue by michaelfig** [Agoric/agoric-sdk#2018]
**Describe the bug**
The current specified behaviour of `@agoric/marshal` is to treat empty objects as `Remotable`.  This is handy for creating object handles which are `===` to themselves when returned to the vat that created them.    However, this convenience comes at a cost of inadvertently creating Presences when only an empty data object is intended.

**To Reproduce**
1. Send an empty object (`{}`) as an argument to a method defined in another vat (via `@agoric/marshal`)
2. `console.log` out the received argument.
3. Note that it displays as a Presence, not just an empty object.

**Expected behavior**
It would be clearer to have empty data objects treated as pass-by-copy.

**Platform Environment**
 - what OS are you using? what version of Node.js? n/a
 - is there anything special/unusual about your platform? n/a
 - what version of the Agoric-SDK are you using? latest master as of 2020-11-13

**Additional context**
See #804 for some discussion

As a migration step, requiring people to explicitly mark empty objects as either `Remotable` or `Data` would be useful to detect up-front when they haven't considered what the object is intended to be.  During this intermediate step, the `Data` marking could be applied automatically to empty objects that are received as pass-by-copy from another vat to prevent false positives (only empty objects constructed from within the vat would need explicit marking).

`Data` can be made optional in a later step when the default for unmarked empty objects is pass-by-copy.

**Comment by erights** [comment]
We thought we could get by treating empty objects as remotables even though they sometimes just represent empty data objects. We missed that there is no semantics for `sameStructure` that works for both interpretations of an empty object.

```js
const x1 = harden({ foo: 1 });
const y1 = harden({ foo: 1 });
safeStructure(x0, y0); // true
```
but
```js
const x0 = harden({});
const y0 = harden({});
safeStructure(x0, y0); // false
```

So we need to make progress on #804 urgently in order to fix this one.
Some progress at https://github.com/Agoric/agoric-sdk/pull/2178

**Comment by warner** [comment]
In today's kernel meeting, we settled on a transition plan:

* Introduce a function, probably named `Data()`, which records the intention to be pass-by-copy. This lives next to the existing `Far()` and `Remotable()` functions.
* Change marshal to warn when it serializes an empty object that is not marked as either Remotable/Far or Data.
* Using the warning as a guide, locate and mark all empty objects as either Far or Data, until the warnings stop.
* Change marshal to throw an error when asked to serialize an unmarked empty object.
* Wait a while, make sure we've caught all the cases.
* Change marshal to treat empty objects as pass-by-copy data, even without the mark.
* Remove the `Data()` markers.

The end/goal state is that pass-by-data is the default (I think we had agreement on this), a `Data` marker is neither needed nor available (I think we mostly had agreement on this), and that all pass-by-reference objects should be marked with Far/Remotable (I'm not sure we had agreement on this).

We certainly want unmarked empty objects to be pass-by-copy, because e.g. @FUDCo 's option-bag use case is best handled that way.

Our current behavior is that unmarked empty objects are pass-by-reference. In one sense this is ok, because they'll arrive as Presences, which look mostly like empty objects, so any code which e.g. enumerates over their properties, or splats them into another object (`x = { a: 1, ...remoteRecordWhichHappensToBeAPresence }`) will accidentally do the right thing. But `sameStructure` treats Presences as distinct, so comparing two records that contain empty records won't work correctly. @dtribble noted that since pass-by-reference empty objects lack type information (because the only way to attach type information is with Far/Remotable), we could work around this temporarily by changing `sameStructure` to treat untyped Presences as equal. But I think we decided that we should execute the plan above instead of using this workaround.

**Comment by warner** [comment]
@erights mentioned that he's happy with this plan

**Comment by michaelfig** [comment]
For future reference, I think it will be simpler just to mark the actual empty-object literals (the exact `{}` occurance, and possibly the return value from `Object.fromEntries(...)` or `Object.create(...)`), but leave them mutable.

That's both more systematic and more tolerant than trying to determine when the object is "ready for delivery", especially since we are currently auto-hardening arguments to eventual sends.

**Comment by erights** [comment]
As an interim step, sure. As a place to land, that would make me very sad.

**Comment by erights** [comment]
Likewise, having our lower level harden objects without even a warning is masking bugs. They should at least warn, and eventually reject unhardened arguments.

**Comment by warner** [comment]
work is happening in https://github.com/Agoric/agoric-sdk/tree/2018-empty-objects

I count 216 warnings in the swingset tests (probably far fewer call sites, each run multiple times).

**Comment by warner** [comment]
I fixed all the occurrences in the swingset tests. I'm now looking at the rest of the tree, currently there are 768 warnings.

**Comment by warner** [comment]
I ran into an old problem with the definition of `Data`: I followed the existing pattern used by `Far/Remotable` (a `WeakMap/WeakSet` defined at the top level of `marshal.js`):

```js
const registeredData = new WeakSet();

function isPassByCopyRecord(val) {
  ...
  const descEntries = Object.entries(Object.getOwnPropertyDescriptors(val));
  if (descEntries.length === 0) {
    // empty non-array objects must be registered with Data or Far/Remotable
    if (!remotableToInterface.has(val) && !registeredData.has(val)) {
      console.log(`--- @@marshal: empty object without Data/Far/Remotable`);
    }
  ...
}

export function Data(obj) {
  harden(obj);
  registeredData.add(obj);
  return obj;
};
```

*but*, our kernel bundle (which includes the copy of `liveslots.js` which is used by all local vats) includes one copy of `marshal.js`, while each vat bundle includes its own (separate) copy of `marshal.js`, and they don't share the same `WeakSet` table. So if vats use `import { Data } from '@agoric/marshal.js'` to get `Data`, their markers won't be recognized by liveslots when it goes to serialize the arguments. This is a form of the "identity discontinuity" problem that made using multiple Realms difficult.

We've seen this problem before, when we first introduced `Remotable`. The question is how to share a singleton with vat code. The wrinkle is that we have code defined in libraries that want to run both inside a vat (for regular operation) and outside of a vat (for unit tests). There are basically three answers:

* 1: import it from a module. Given our current `bundleSource`/`rollup`/`importBundle` approach, this has the identity-discontinuity problem, which could be addressed by:
  * tell `rollup` that the module in question is an "external", so `rollup` will leave a `require()` in place, and then provide a special `require` global to vats, which is wired to a fake module that returns the singleton
  * add the singleton as a global, change the module to only define the singleton if it is "first on the scene", deferring to the global if it already exists
* 2: fetch it from a global
* 3: pass it as an argument (the `vatPowers` to `buildRootObject`), threaded through the entire call stack down to the code that needs it

Importing from a module is the most ergonomic answer. Fetching it from a global is easy to code against, but unit tests would need some externally-supplied shim to inject the global first. Passing it as as argument is really painful, because when `buildRootObject` calls A, A calls B, B calls C, and C wants to use `Data`, all of A/B/C must include `Data` as an argument, which bloats the function signatures and is awfully hard to maintain.

We currently manage `harden()` as a global, such that all code using `harden` is declared to require a "SES environment". Unit tests achieve this by importing `@agoric/install-ses` before any other imports, while vats bundles are evaluated in a SES `Compartment` which has a `harden` global as the SES spec defines.

The `Remotable` function is currently managed as an argument, passed into `vatPowers`. It is used in `captp`, but no other user-level code like ERTP or Zoe.

The recently-added `Far` function (which is implemented in terms of `Remotable`) is managed as an imported module, however this suffers from the identity-discontinutity problem, so I don't think it's actually functional in real vats right now (cc @erights).

## Marking with a Symbol instead of registering in a WeakMap

A fourth approach would be to mark objects as `Data` or `Remotable` by adding a non-enumerable property whose "name" is a registered Symbol. For `Data`, the property's value could just be `true`, while for `Remotable` it could be the interface name.

This would address the identity discontinuity problem because any two acts of registering the same name will yield Symbols that share an object identity: `Symbol.for('@agoric/marshal:Data') === Symbol.for('@agoric/marshal:Data')`, no matter what bundles or acts of evaluation contain the two sides. The left side can be the copy of `marshal.js` that is imported by liveslots in the kernel bundle, while the right side could be the other copy of `marshal.js` that is imported into a vat bundle.

(*Unregistered* Symbols, in contrast, are unforgeable: `Symbol('data') !== Symbol('data')`, and the string is merely a diagnostic description, rather than an identity. Using an unregistered Symbol would cause the same problem that the unforgeable WeakMap/WeakSet has)

(Yet another option that @FUDCo pointed out would be to make the WeakMap a global. This wouldn't cause any direct authority problems, because WeakMaps are unenumerable, but could still provide an illicit communication channel because widely-held objects like Object and Number could be added/removed by otherwise mutually-incommunicado code)

By making the marker property be unenumerable, casual users of the object won't be able to tell that it is marked (they would have to use `Object.getOwnProperties` or similar to see the symbol property). Also, because our serialization layer uses `JSON.stringify`, the serialized form will ignore the marker property too, so remote users won't be able to tell either (even if they *do* look deeper).

Note that this makes `Data` and `Far`/`Remotable` forgeable in a way they weren't previously: not only could code use `Object.getOwnPropertySymbols` to extract the Data symbol and attach it to a different object, but it can also just synthesize a new copy of the symbol itself, with `Symbol.for`. But we don't need these markers to be unforgeable, so this isn't an issue.

One other difference is that, since we're adding a property, we cannot apply `Data` to an object that is already hardened. As @erights pointed out, this is actually a feature: if I create an object as pass-by-reference, you should not be able to cause my object to be treated as pass-by-copy. I can prevent this by merely hardening the object ahead of time, which I ought to be doing anyways. Most code will apply the `Data` or `Far` marker immediately surrounding the object literal:

```js
const empty = Data({});
const handle = Far('interface name', {});
```

@FUDCo 's "cumulative option bag" use-case defers applying the marker until after the bag is full, but before the object is revealed outside the construction function:

```js
const options = {};
for (let opt of argv) {
  ...
  options[name] = value;
}
return Data(options);
```

So my new plan is to have `marshal.js` contain something like:

```js
const DataSymbol = Symbol.for('@agoric/marshal:Data');
function Data(obj) {
  Object.defineProperty(obj, DataSymbol, { enumerable: false, value: true });
  return harden(obj);
}
```

and change the `passStyleOf` code to use `obj[DataSymbol]` (or maybe `Object.getOwnPropertyDescriptor(obj, DataSymbol).value`) to sense it.

I think I need to change `Remotable` and `Far` to do the same thing (using the interface name instead of `true` as the property's value), because I think `Far` is currently broken.

@michaelfig I wanted to check this plan with you, I know you and @erights circled around this issue a lot, and settled on the current `vatPowers.Remotable` approach with good reason. If I'm missing something, please let me know.

**Comment by warner** [comment]
It looks like I need to change the way `isPassByCopyRecord` works, to ignore non-enumerable properties (it currently looks at properties with `Object.getOwnPropertyDescriptors`, which includes the non-enumerable ones), and to tolerate Symbol-named properties (it has code which checks for these, but I think it's broken/bypassed, because it uses `Object.entries` around the descriptors, which ignores Symbols, but then the actual marshalling code uses `Reflect.ownKeys().sort()`, which fails because Symbols can't be auto-stringified, so they can't be sorted).

**Comment by erights** [comment]
> to ignore non-enumerable properties

>  Symbol-named properties

No, that's not at all what I have in mind. We need to fix the bug you identified in `isPassByCopyRecord` so it is *less* tolerant. We should tolerate only the symbol properties we're introducing to communicate with marshal -- marking something remotable or data. Any other symbols, and any other non-enumerable properties, must cause rejection. These special properties should not be serialized as properties. Rather, their only affect on serialization is how the object is classified, and thus how or whether the other properties are serialized.

**Comment by warner** [comment]
Got it. Out of curiosity, what's the rationale for rejecting (rather than ignoring) those other properties? Is it to ensure that the round-tripped object is the same as the original (specifically that `Object.getOwnPropertyDescriptors()` of the copy is deep-equal to that of the original)?

That reminds me, I think we need to also change the *deserializer* to add the same Data marking as the original. We want to make sure that an empty (Data) object can be passed into a method as an argument, and then passed back out again (as the return value, or in an argument of an outbound message), without the code needing to add its own `Data(obj)` in the middle.

**Comment by erights** [comment]
> That reminds me, I think we need to also change the deserializer to add the same Data marking as the original. 

Yes. In the same way that the unserializer adds the same Remotable marking as the original. Data does not have the same third party problem since everyone is equally authoritative over their own copy, so the way we're currently restoring the Remotable marking will actually be correct for Data.

**Comment by erights** [comment]
> Got it. Out of curiosity, what's the rationale for rejecting (rather than ignoring) those other properties? Is it to ensure that the round-tripped object is the same as the original (specifically that Object.getOwnPropertyDescriptors() of the copy is deep-equal to that of the original)?

* The semantics of our distributed object system is in terms of a much simpler base semantics, and a closer-to-language-independent one, than JS itself. 

* Our own code processing Data objects we receive sometimes uses primitives that enumerate all own properties. Sometimes uses primitives that enumerate only enumerable own properties. "..." both for expressions and destructuring patterns is one of the most convenient ways of manipulating objects, but operates only on enumerable own properties. Our code currently assumes all of these agree. I would rather make that code correct by removing the difference than force it to deal with the complexities of these operations being different. I would also like to relieve our users of that hazard.

**Comment by erights** [comment]
> without the code needing to add its own Data(obj) in the middle.

Yes. But the code in the middle can't anyway because the object is hardened.

**Comment by warner** [comment]
@erights FYI, I found that we're currently adding a `Symbol.toStringTag` to a few objects, which the marshal code needs to tolerate (ignore, do not serialize). It occurs here:

https://github.com/Agoric/agoric-sdk/blob/814ebd2f0b74e1a9d50c8b0f5f91ea31cd6e3b78/packages/SwingSet/src/weakref.js#L48-L53

Soon, liveslots will need access to `WeakRef` and `FinalizationRegistry` to perform its semi-determinstic GC work. This work is half-complete, but liveslots is currently passed both, and there are a few tests to exercise their use. As a measure of compatibility with Node.js v12 (which does not provide either unless you pass an unstable CLI `--flag`), `src/weakref.js` defines a fake `WeakRef` (which actually holds a strong ref) when it's not otherwise available. To make this fake `WeakRef` behave mostly like the real one, I found I needed to add a `Symbol.toStringTag` to the instances.

(Hm, apparently I wrote the first version of the fake WeakRef, but then you made a cleanup pass, so `toStringTag` might be in code that you wrote)

It's not a big deal to have `marshal` ignore this particular property when prohibiting other Symbol-named properties (the same code will ignore our Data/Far marker properties), but if we don't like this, we could either find a different way to build the fake WeakRef, or drop compatibility with the older Node.js and always pass a real WeakRef in.

**Comment by warner** [comment]
@erights I was completely wrong about the source of the `toStringTag`. It's actually coming from `marshal.js` itself, when it creates a `Remotable`: a new prototype is inserted between the Remotable and its original prototype, the inserted one includes both `toString` and `Symbol.toStringTag`, and `mustPassByRemote()` checks that the prototype is `passByRemote` too:

https://github.com/Agoric/agoric-sdk/blob/e628351668f10dfdb868631329944eedd03730f1/packages/marshal/src/marshal.js#L903-L910

I was tricked by the fact that the failure only showed up in the SwingSet tests: apparently the `marshal` tests don't ever subject a `Remotable` to `mustPassByPresence` (it can be passed to `serialize` without complaint, but `mustPassByPresence`/`mustPassByRemote` complains). Swingset does do that: liveslots tests the return value of `buildRootObject` against `mustPassByPresence` to fail fast if the user code does something wrong.

I don't understand the "is the prototype `passByRemote`" check. I also don't understand why `serialize` tolerate this: I see comments in `mustPassByRemote` that suggest it wouldn't be called if the object is registered in the Remotable table, but I'm not sure what might be doing such a check, or whether the check is unnecessary in that situation.

My current plan is to update `test-marshal.js` to include a `mustPassByPresence(Far(..))`, to trigger this situation, and then make serialization tolerate (ignore, do not serialize) the `Symbol.toStringTag` property on both pass-by-reference (which is causing the problem) and pass-by-copy (which is not, currently, but it seems like a good idea to keep them in sync).

**Comment by warner** [comment]
that first PR merely adds tests of the existing behavior, without changing that behavior

**Comment by warner** [comment]
I should record the fact that our current plan is to make `Far`/`Remotable` mandatory for all pass-by-reference objects. When we're done with the transition process, it will be an error to serialize e.g. `harden({ foo: () => 0 })`. The rules will be:

* everything must be hardened
* if it's marked with Far/Remotable, it will be pass-by-reference
  * for now, it may only contain methods. When #2069 (auxdata) is implemented, it may contain data too
* else it's pass-by-copy, and must only contain data
  * if/when we implement pass-by-reference of bare Functions, someday, it may contain functions too
* there is no Data() marker

**Comment by erights** [comment]
> My current plan is to update test-marshal.js to include a mustPassByPresence(Far(..)), to trigger this situation, and then make serialization tolerate (ignore, do not serialize) the Symbol.toStringTag property on both pass-by-reference (which is causing the problem) and pass-by-copy (which is not, currently, but it seems like a good idea to keep them in sync).

Let's talk about that before you do this. I won't have time to investigate until tomorrow (Thurs) afternoon. But from your description I think marshal's `serialize`, `unserialize` and `passStyleOf` behaviors are correct as is, but the `must*` methods are wrong.

**Comment by warner** [comment]
I asked @dtribble at lunch, and he said he's ok with the idea that all pass-by-reference objects must be marked with e.g. `Far`. His main interest was that all such objects have an interface name/type assigned to them, which either means we call a marker (`Far('interfaceName', { ...methods })`), or maybe have a magic property for the interface name (`harden({ INTERFACENAME: 'interfaceName', ...methods })`), and I think `Far` winds up being faster to type.

@michaelfig and I went through the issues, both for this and the increasingly-entangled #2069 auxilliary data, and I think I have a better understanding of the issues now:

* we want Remotables to be displayed as such by `console.log` and a debugger, along with their declared interface name
  * but we don't want to add any surprising properties on them, which might confuse code using `Object.getOwnPropertyNames`
* we want the Presence that pops out the other side to have similar debugging info
  * it's even more important to avoid surprising properties on that Presence, because Presences are supposed to be empty objects (don't imply the existence of any local data)
* we need Presences to be recognized as Remotable too (liveslots, which is responsible for creating Presences inside the `convertSlotToVal()` function it gives to marshal, uses `HandledPromise` code to make the Presence, and then marks it by calling `Remotable`)
  * extra properties on a Presence could interfere with that
* finally, marshal's `mustPassByReference` performs a check that the perhaps-Remotable object's `.prototype` itself qualifies as `mustPassByReference`. I still don't entirely understand this, but I think it might be in anticipation of Remotables including the identity of the prototype in their serialized form, allowing the receiving end to set up a prototype chain that mimics the one on the sender.
* `mustPassByReference` was not expected to accept any object returned by `Remotable()`, because `passStyleOf` started by looking up the object in the WeakMap which `Remotable` populates, so it would never get as far as the `mustPassByReference` check. Liveslots was expecting more out of `mustPassByReference` than it was designed to do.

Previously, the design that met all these contraints was to insert a new prototype above the Remotable, which contains a `toString` and a `Symbol.toStringTag`. The debugger (and console.log) will see these when printing the object (satisfying the debuggability requirement), but because they're on the prototype instead of the real object, serialization and other users will *not* (satisfying the "no surprising properties" requirement).

@michaelfig and I decided that, instead of a prototype, we could just put an enumerable Symbol-named property on the object itself, and use the interface name as the value. He suggested scoping it like `Symbol.for('agoric.Far')` or similar, although @dtribble then pointed out that `endo.Far` might be better (if we treat eventual-send and marshalling as part of the Endo project, whose scope is broader than Agoric, attn @kriskowal for thoughts). By using a Symbol, most casual users won't notice it (Object.getOwnPropertyNames ignores symbols). And `console.log` (and the debugger) will include both the symbol name and the value when displaying the object, which includes the interface name, meeting the debuggability requirement.

So the current plan is:

* change `Remotable` to merely:
  * add `val[Symbol.for('agoric.Far')] = interfaceName`
  * `harden()`
  * (and maybe do some fail-early checking that the object meets the `mustPassByReference` requirements, like no unenumerated properties, or Symbol-named properties other than `agoric.Far`)
* change `passStyleOf` to start with an early check for `val[farSymbol]`, and if present, return "pass-by-reference" right away
  * during the transition `passStyleOf` will then check for `val[dataSymbol]`, which will be unenumerated
  * if neither symbol is seen, examine the properties to decide between pass-by-reference and pass-by-data
  * in either case, before actually serializing, we must assert that all the other requirements are met: no unexpected symbol/unenumerable properties, hardened, the right kind of prototype, etc
* `mustPassByReference` will *not* tolerate `Symbol.toStringTag` or `toString`, we'll no longer be adding those to anything
* liveslots could keep using `mustPassByReference` aka `mustPassByPresence` (which will now accept Remotables, because `Remotable()` no longer inserts a prototype), but could also switch to `passStyleOf(obj) === 'pass-by-reference'` (assuming it makes all the right property checks) which might express what it cares about more precisely
* `getInterfaceFor` changes to merely look for the `Symbol.for('agoric.Far')` property on the object, rather than looking in a WeakMap
* `marshal` will no longer have any top-level `WeakMap`s
* `passStyleOf`, `Remotable`,  `getInterfaceFor`, and `Far` can all be plain exports of the `marshal` module, rather than needing to be passed through `vatPowers`, since they won't need to share a `WeakMap`

For the interim `Data` marker:
* `Data(obj)` will work by adding a non-enumerable Symbol-named property like `Symbol.for('agoric.Data')` to the object
  * I think it should also harden() the object, @michaelfig points out that `Data()` is temporary and so we shouldn't take on any more pain than we really need to. I'm inclined to `s/harden({})/Data({})/g` instead of making it be `harden(Data({}))`. We didn't come to agreement on that point.
* `passStyleOf` will first look for the marker, if that is not found it will look at the properties, and all-non-function properties mean the object gets classified as pass-by-copy
* `mustPassByCopy` tolerates the Data marker symbol, but rejects any other Symbol-named properties or unenumerated properties. The serializer can thus only pay attention to enumerable properties, or enumerable string-named properties, as before.
* to make sure deserialized empty objects can be re-serialized without complaint, `fullRevive()` (in the last `else` clause, just before calling `ibidTable.finish()`) *must* add the Data marker symbol to the object it creates, if that object has no property names.

**Comment by warner** [comment]
I just got off the phone with @erights , who pointed out that we can't really afford to use an enumerable Symbol on the Remotable as a marker, because that will be visible to the `...` operator (as in `obj2 = { newprop: newvalue, ...obj1 }`). And if it's *non*-enumerable, it doesn't show up in the debugging info (at least under the Node.js repl):

```
> o = Object.create(Object.prototype, { [Symbol.for("tag")]: { enumerable: false, value: 1 } })
{}
> o
{}
> o = Object.create(Object.prototype, { [Symbol.for("tag")]: { enumerable: true, value: 1 } })
{ [Symbol(tag)]: 1 }
```

So our new plan is to stick with the inserted prototype in `Remotable`, but add our new `Symbol.for('Far')` *on the prototype*, rather than on the child object. And `passStyleOf` will do a simple `obj[Symbol.for('Far')]` check to look for the marker, which follows the inheritance chain.

And since the marker is going onto the prototype, we could leave it enumerable, rather than making it non-enumerable: the only iterator which pays attention to properties of the prototype is `for..in`, which we never use (and expect to add lint rules to make sure nobody else uses it either). I have to look more closely at the current implementation of `Remotable`; I'll make the Far marker do the same thing as the `toString` and `Symbol.toStringTag`, which might actually be non-enumerable right now (I never remember if `Object.create` defaults to enumerable or non).

@erights is taking over the 2018-reject-symbols branch and PR #2331, and is checking all places where he used `Object.entries` in light of the surprising awareness that it only returns string-named properties.

One other thing I learned is that pass-by-reference needs to tolerate (and ignore) `Symbol.asyncIterator` on the object. `marshal` needs to be able to serialize this symbol as a piece of data (and only that symbol, although we will add other "unregistered but well-known" symbols in teh future). Pass-by-reference will reject any other symbols, and no symbols get included in the serialized form, but we should not reject an object as pass-by-reference just because it (locally) can be used as an async iterator.

**Comment by warner** [comment]
I've got `Far` and `Data` added, and I have PRs up to use them in swingset, ERTP, and captp. I'm working on Zoe now. For reference, the process is:

* run your tests, doublecheck they work first
* uncomment the two asserts in `marshal.js` (look for `@@marshal:` and "empty object without Data/FarRemotable" and "pass-by-ref object without Far/Remotable", currently at lines 269 and 504)
* run your tests again, note where those assertions are triggered
* for each file implicated:
  * add `import { Far, Data } from '@agoric/marshal';`
  * replace `harden({ methods..})` with `Far('interface name', { methods.. })`
  * replace `harden({})` with either `Data({})` or `Far('interface name', {})` as appropriate
  * repeat until the asserts stop firing and the test passes

Another way to direct the search is:
* find all occurences of `buildRootObject`, which means that file/library creates a vat
* change the return value of `buildRootObject` must be marked as `Far`. I've been using an interface name of `'root'` for this.
* walk through the rest of that file to find all other return values that look pass-by-reference (`harden` with methods), and mark that as Far too
* find all imports of that file, repeat

If the assertion fires during a non-swingset test, you might get a stack trace that shows you a line number shortly after the creation of the object that needs to be marked. Assertions that fire during swingset-based tests are less likely to guide you.

**Comment by warner** [comment]
Current status: I have PRs landed or in-flight (#2564 #2565 #2566) for adding Far/Data to all agoric-sdk packages, except cosmic-swingset (#2567) which needs @michaelfig 's help.

In parallel with that, I think we need to update our dapps that live outside the agoric-sdk monorepo.

Once we've updated all those, the next step is either to enable the `marshal.js` unmarked-empty-object warning `console.log`, or jump ahead and enable the exception too. Enabling the exception will break any third-party dapps which aren't following the new guidelines yet. In practice, the `Far`-marking requirement seems to require more code changes than the `Data({})` requirement. Do we have a good place to contact those authors and tell them about the changes?

We must decide how long we'll wait in this "throw errors for unmarked empty object" state. The idea is to catch any last cases of unmarked `{}` meaning pass-by-reference. Then we can remove the `Data({})` requirement and rewrite our code to use `harden({})` instead. And then after some *additional* period, we can remove `Data()`, or make it a no-op.

**Comment by katelynsills** [comment]
> I think we need to update our dapps that live outside the agoric-sdk monorepo.

Towards that end, here's a checklist of the dapps that can be copied for an issue: https://github.com/Agoric/agoric-sdk/issues/2555

> Do we have a good place to contact those authors and tell them about the changes?

We don't right now, although we can post in Discord as a general mechanism.

**Comment by warner** [comment]
I talked with @michaelfig just now, and we think:

* marking all pass-by-reference objects with `Far` is a worthy goal, and we should do it, but is not necessary to achieve the original purpose of treating empty objects as data
  * so let's pursue that in a separate ticket, on a separate timeline
* the task of tracking down everything which is meant to be data but might be empty (and thus marking it) is tedious and annoying
  * and is only necessary if we want to use an exception to catch the places where we've missed `harden({})` that's meant to be data instead of Far
  * so let's not do that

Our current plan is to:
* *grep* for `{}` (especially `harden({})`), one package at a time, replacing some with `Far(iface, {})`, and then deciding that we've probably found all of them
  * including the dapps
* make sure tests pass, but don't turn on any asserts in marshal
  * enabling the unmarked-empty-object warning is interesting, and if we see it trigger in a surprising place we should investigate, but we aren't trying to suppress all of them
* change `@agoric/store` to reject keys which are pass-by-data empty objects (assert that they are pass-by-ref)
  * we think this might catch most leftover cases, and it's pretty clearly a mistake
  * make sure tests still pass
* change marshal to accept hardened empty objects as data
* make sure tests still pass
* remove Data

**Comment by erights** [comment]
For the development stage we're in, at this point, I think this is a fine plan.

**Comment by warner** [comment]
Current status: `harden({})` is now pass-by-copy, `Far()` is recommended but not required (no warnings are produced, but you could uncomment lines in `marshal.js` to enable warnings and/or hard errors), `Data()` is still available but not required.

The next step is to strip out Data  (#2632). Then establish a different ticket (#2634) for promoting/requiring the use of Far.

**Comment by erights** [comment]
See also https://github.com/Agoric/agoric-sdk/pull/3558
